### PR TITLE
KAFKA-12650: fix NPE in InternalTopicManagerTest

### PR DESCRIPTION
**Pull request by mjsax**
The NPE arrises, if we cannot setup topics quickly enough, and run into a timeout -- for this case, we try to call `admit.delete` (note that `admin` is mocked) that returns `null`.

However, we should never actually "abort" the internal topic creation. -- Using `MockTime` instead of `SystemTime` should avoid this issue, as time won't advance at all. Or would there be any concern with regard to running forever? An alternative approach could be, to increase `max.poll.interval.ms` that defines the timeout.

Call for review @bbejeck @ableegoldman

### CODE REVIEW

**Comment by ableegoldman** [comment]
Personally I would slightly prefer to just increase the max.poll.interval if that's what defines the timeout (although, why would the max poll interval impact a timeout related to the admin client? Or do I not understand where this timeout is ultimately coming from)

But we should probably try to actually handle the `null` even so, and fail with a more useful error message in case we ever do hit this again. Or if we can't handle the null directly, maybe use a not-nice mock so it'll fail rather than just returning null and hitting an NPE

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java]: In test setup, we reduce max.poll to 100ms, to increasing it to 10 sec for this test should be sufficient.

**Comment by mjsax** [comment]
Updated the PR.

> (although, why would the max poll interval impact a timeout related to the admin client? Or do I not understand where this timeout is ultimately coming from)

We use `InternalTopicManager` in the consumer group leader `assign()` callback to create internal topics. If we get errors back from admin client, we retry. To bound the retries, we use `max.poll.interval.ms / 2` because `max.poll.interval.ms` defines an upper bound before `assign()` must finish (otherwise the leader gets kicked out of the group). (Cf. KIP-572).

> But we should probably try to actually handle the null even so, and fail with a more useful error message in case we ever do hit this again. 

`admin.deleteTopic` would never return `null` in production, and thus handling `null` within `InternalTopicManager` for this case seems not to make sense.

**Review by showuon — COMMENTED** [review]
@mjsax , thanks for the explanation. Make sense! There are check style errors, please update it. Otherwise, LGTM.
```
[2021-04-13T02ː52ː28.842Z] [ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-10529/streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.javaː67ː15: Unused import - org.easymock.EasyMock.anyObject. [UnusedImports]
[2021-04-13T02ː52ː28.842Z] [ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-10529/streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.javaː69ː15: Unused import - org.hamcrest.CoreMatchers.any. [UnusedImports]

```

**Comment by ableegoldman** [comment]
> admin.deleteTopic would never return null in production, and thus handling null within InternalTopicManager for this case seems not to make sense.

Agreed, making it a strict mock seems sufficient for avoiding the NPE then. Thanks

**Review by ableegoldman — APPROVED** [review]
LGTM, modulo the checkstyle failures. As long as @cadonna  has no objections as the author of this test then feel free to merge once it's passing

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java]: I think a default mock would be enough to get a better error message. Strict mocks also check the order of the calls which is not needed here. 

I would also pass `new MockTime(1)` instead of increasing the timeout, because the test retrieves the time twice. `new MockTime(1)` would progress time 1 ms each time. Hence, we would be within the 100 ms of the global `max.poll.interval.ms` and since we control the time progression the test would be more robust than by increasing `max.poll.interval.ms` to a higher value. With `new MockTime(1)`, we also avoid running the test indefinitely in case of a mistake in the production code, since after 50 time retrievals, the test would run into a timeout and the admin client mock would throw the unexpected method call error. Probably, it would be even better to use something like `new MockTime((Integer) config.get(StreamsConfig.consumerPrefix(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG)) / 3)`.

Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java]: After this PR is merged, I will open a PR that changes also the other usages of `Time.SYSTEM` and `createNiceMock()` since I think that would make all tests more robust. I do not do it now because I am still working on that code for KIP-698 and I do not want to run into too many merge conflicts.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java]: Could you please revert this change since I am still working on this code for KIP-698 and I do not want to run into merge conflicts?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java]: Cool. Works for me.

**Comment by mjsax** [comment]
Updated the PR.

**Review by cadonna — APPROVED** [review]
@mjsax Thanks!

LGTM!

**Comment by cadonna** [comment]
3 unrelated test failures.

```
Build / JDK 15 and Scala 2.13 / kafka.server.RaftClusterTest.testCreateClusterAndCreateListDeleteTopic()
Build / JDK 15 and Scala 2.13 / kafka.server.RaftClusterTest.testCreateClusterAndCreateListDeleteTopic()
Build / JDK 15 and Scala 2.13 / org.apache.kafka.streams.integration.KTableKTableForeignKeyInnerJoinMultiIntegrationTest.shouldInnerJoinMultiPartitionQueryable
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java
Comment (chia7712) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:156]: Is this a necessary change? If so, should we change all similar code in this test class?

**Review by chia7712 — APPROVED** [review]
@mjsax thanks for this fix.

loop test with this patch. NPE does not happens again. +1

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:156]: See https://github.com/apache/kafka/pull/10529#discussion_r612291115

Comment (chia7712) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:156]: @cadonna thanks!

**Pull request merged**
